Add tests for DaemonSet lvmd

bells17 opened the pull request:
ref: #195
old PR is here: #265

Hi, I added tests and doc for DaemonSet lvmd.
The details are as follows:

* Added tests for DaemonSet lvmd(using minikube in this tests because lvmd container can't use on [kind](https://kind.sigs.k8s.io/) cluster).
* Added github action jobs for running DaemonSet lvmd tests.
* Added the description about how to deploy DaemonSet lvmd.

bells17 commented on e2e/node_test.go:
```diff
@@ -37,7 +37,12 @@ func testNode() {
+				count = 1
+			}
+
+			if len(podlist.Items) != count {
```
https://github.com/topolvm/topolvm/pull/265/files#r595008742
> I assume this test checks topolvm-node should be deployed on all nodes,
so check if pod number equals worker node number.

when using kind, I need to exclude control plane nodes because control plane nodes don't deploy non control plane pods to themselves.

```
$ kubectl get node --selector=node-role.kubernetes.io/control-plane!=''
NAME                  STATUS   ROLES    AGE     VERSION
topolvm-e2e-worker    Ready    <none>   5m50s   v1.20.2
topolvm-e2e-worker2   Ready    <none>   5m50s   v1.20.2
topolvm-e2e-worker3   Ready    <none>   5m50s   v1.20.2
```

But, on the other hand, when using minikube(bare metal), if I exclude control plane nodes, I can't find the worker node because the worker node is also a control plane node.

```
$ kubectl get node --selector=node-role.kubernetes.io/control-plane!=''
No resources found
```

So I think it is difficult to compare the counts of pods and nodes dynamically.

toshipp replied:
Thank you for your explanation.
I got it.

llamerada-jp commented on deploy/manifests/lvmd/lvmd.yaml:
```diff
@@ -0,0 +1,70 @@
+kind: ServiceAccount
+metadata:
+  namespace: topolvm-system
+  name: topolvm-lvmd
```
```suggestion
  name: lvmd
```

The ServiceAccount name for node is `node` and controller is `controller`. `lvmd` is better if we follow it.

bells17 replied:
fixed  it.
https://github.com/topolvm/topolvm/pull/285/commits/2f348e042a1885c1b417cd68e2f8f955893fcc26

llamerada-jp commented on e2e/Makefile:
```diff
@@ -151,4 +159,69 @@ setup:
 	curl -o $(BINDIR)/kubectl -sfL https://storage.googleapis.com/kubernetes-release/release/v$(KUBERNETES_VERSION)/bin/linux/amd64/kubectl
 	chmod a+x $(BINDIR)/kubectl $(BINDIR)/kind
 
-.PHONY: launch-kind shutdown-kind start-lvmd stop-lvmd test clean setup
```
Could you rewrite `PHONY` for each target like below.

```
.PHONY: clean
clean:
    ~~~

.PHONY: setup
setup:
    ~~~~
```

bells17 replied:
fixed it.

llamerada-jp commented on deploy/README.md:
```diff
@@ -45,6 +45,50 @@ To setup `lvmd`:
 
 3. Install `lvmd` and `lvmd.service`, then start the service.
 
+### OPTIONAL: Launch lvmd with DaemonSet
```
Could you describe the top of the section that lvmd can be deployed as a daemon or using daemonset?

bells17 replied:
Excuse me. I updated the lvmd description in the below commit:
https://github.com/topolvm/topolvm/pull/285/commits/aeef54c190f709d11c3d517701eaec6ae6f535a7
Is the issue that you commented on resolved?

llamerada-jp replied:
Yes, it's fine.

llamerada-jp commented on deploy/README.md:
```diff
@@ -45,6 +45,50 @@ To setup `lvmd`:
+
+Also, you can launch [lvmd][] with DaemonSet.
+
+Notice: The lvmd container uses nsenter to running some lvm commands(like `lvcreate`) as a host process, so you can't launch lvmd with DaemonSet when you're using [kind](https://kind.sigs.k8s.io/).
```
```suggestion
Notice: The lvmd container uses `nsenter` to run some lvm commands(like `lvcreate`) as a host process, so you can't launch lvmd with DaemonSet when you're using [kind](https://kind.sigs.k8s.io/).
```

llamerada-jp commented on deploy/README.md:
```diff
@@ -45,6 +45,50 @@ To setup `lvmd`:
+
+To setup `lvmd` with Daemonset:
+
+1. Prepare LVM volume groups.  A non-empty volume group can be used because LV names wouldn't conflict.
```
```suggestion
1. Prepare LVM volume groups in the host. A non-empty volume group can be used because LV names wouldn't conflict.
```

llamerada-jp commented on e2e/Makefile:
```diff
@@ -151,4 +159,69 @@ setup:
+daemonset-lvmd/setup-minikube:
+	mkdir -p $(BINDIR)
+	$(SUDO) apt-get update
+	$(SUDO) apt-get install -y conntrack
```
```suggestion
	DEBIAN_FRONTEND=noninteractive $(SUDO) apt-get install -y --no-install-recommends conntrack
```

llamerada-jp commented on e2e/Makefile:
```diff
@@ -151,4 +159,69 @@ setup:
+daemonset-lvmd/delete-minikube:
+	$(BINDIR)/minikube delete || true
+
+# Set scheduler configs manually because minikube can't edit scheduler configs
```
```suggestion
# Set scheduler configs manually because minikube can't edit scheduler configs.
```

llamerada-jp commented on e2e/suite_test.go:
```diff
@@ -68,19 +68,42 @@ func waitKindnet() error {
+	return os.Getenv("DAEMONSET_LVMD") != ""
+}
+
+func skipTestIfNeeded() {
```
```suggestion
func skipTestIfDaemonsetLvmd() {
```

`Needed` is too abstract.

toshipp replied:
Would you add comments before using this function?
It may be `Skip because this test requires multiple nodes but there is just one node in daemonset lvmd test environment.`

bells17 replied:
renamed and added coomments.

llamerada-jp commented on e2e/suite_test.go:
```diff
@@ -68,19 +68,42 @@ func waitKindnet() error {
+
+	var nodes corev1.NodeList
+	err = json.Unmarshal(stdout, &nodes)
+	Expect(err).ShouldNot(HaveOccurred())
+	return nodes.Items[0].Name
```
```suggestion
	Expect(err).ShouldNot(HaveOccurred())
	Expect(nodes.Items).Should(HaveLen(1))
	return nodes.Items[0].Name
```

llamerada-jp commented on e2e/publish_test.go:
```diff
@@ -191,16 +200,21 @@ spec:
+			nodeName = getDaemonsetLvmdNodeName()
+		}
+
+		lvYaml := []byte(fmt.Sprintf(`apiVersion: topolvm.cybozu.com/v1
```
Could you replace it using embed.
https://golang.org/pkg/embed/

llamerada-jp commented on e2e/publish_test.go:
```diff
@@ -88,17 +92,22 @@ func testPublishVolume() {
+
 		By("creating a logical volume resource")
-		lvYaml := []byte(`apiVersion: topolvm.cybozu.com/v1
+		lvYaml := []byte(fmt.Sprintf(`apiVersion: topolvm.cybozu.com/v1
```
Could you replace it using embed.
https://golang.org/pkg/embed/

toshipp replied:
I described our context.
As of go 1.16, it introduces embed feature, and we prefer using it instead of writing content directly to source code.
We tend to create `testdata` directory, put `some.yaml` in there, and use embed to read it.
https://github.com/cybozu-go/neco-apps/blob/main/test/storage_test.go#L21-L22

toshipp replied:
BTW, this rewrite is not related to this PR and there are many candidates.
@llamerada-jp
I think the work should be separated into another PR.
How do you think?

llamerada-jp replied:
Sure. It is better to make another PR for replacing with embed.

toshipp commented on e2e/cleanup_test.go:
```diff
@@ -19,12 +19,14 @@ const cleanupTest = "cleanup-test"
 
 func testCleanup() {
 	It("should create cleanup-test namespace", func() {
+		skipTestIfNeeded()
```
I consider it is better to use just one `skipTestIfNeeded` to skip the entire test.
Can we wrap this test with new `It` and call `skipTestIfNeeded` in the first line of it?

bells17 commented on e2e/Makefile:
```diff
@@ -151,4 +159,69 @@ setup:
-.PHONY: launch-kind shutdown-kind start-lvmd stop-lvmd test clean setup
+daemonset-lvmd/create-vg:
+	mkdir -p build $(BACKING_STORE)
+	if [ -f $(BACKING_STORE)/backing_store_lvmd* ]; then $(MAKE) $(@D)/remove-vg; fi
```
fixed it.

toshipp commented:
@bells17 
This is redundant If you miss my comment.
In `clean_test.go`, all `It` blocks are skipped, so can we group them by `context` or `it` and use just one skipping function instead of using a bunch of skips?

llamerada-jp commented on deploy/manifests/lvmd/lvmd-configmap.yaml:
```diff
@@ -0,0 +1,14 @@
+kind: ConfigMap
+metadata:
+  namespace: topolvm-system
+  name: topolvm-lvmd
```
Sorry, my point was incorrect. The configuration for lvmd with daemon is saved in `deploy/lvmd-config/lvmd.yaml`.
So, it is no problem to put the Daemonset lvmd configurations under `deploy` folder.

satoru-takeuchi commented on deploy/README.md:
```diff
@@ -45,6 +47,50 @@ To setup `lvmd`:
+        socket-name: /run/topolvm/lvmd.sock
+        device-classes:
+          - name: ssd
+            volume-group: myvg1 # Change this value to your LV name.
```
s/your LV name/your VG name/

satoru-takeuchi reviewed (commented):
Just one minor change request.

bells17 commented:
> This is redundant If you miss my comment.
In clean_test.go, all It blocks are skipped, so can we group them by context or it and use just one skipping function instead of using a bunch of skips?

Sorry, I missed your comment. And I fixed this point like follow: https://github.com/topolvm/topolvm/pull/285/commits/a51043c8fb645d80d35200fc15e44d7071a4ba0f

I tried 2 pattern skips to fix codes:

* nested it

  ```go
  Context("context", testHook)
    It("parent it", func() {
      skipIfDaemonsetLvmd()
      It("child it", func() {
        ..
      })
    })
  })
  ```

* use skip in context

  ```go
  Context("context", testHook)
    skipIfDaemonsetLvmd()
    It("it", func() {
      ..
    })
  })
  ```

However, I got an error with these patterns, so I modified the code using `BeforeEach`.
ref: https://github.com/kubernetes-csi/csi-test/blob/v4.1.0/pkg/sanity/controller.go#L158-L162

toshipp reviewed (approved):
LGTM.
waiting @llamerada-jp' review

llamerada-jp reviewed (approved):
LGTM
